karapace/kafka_rest_apis/__init__.py

```diff
@@ -297,19 +524,26 @@ async def cluster_metadata(self, topics: Optional[List[str]] = None) -> dict:
             if self._metadata_birth is None or time.monotonic() - self._metadata_birth > self.metadata_max_age:
                 self._cluster_metadata = None
 
-            if self._cluster_metadata and topics is None:
-                return self._cluster_metadata
+            if self._cluster_metadata:
+                # Return from metadata only if all queried topics have cached metadata
+                if topics is None:
+                    if self._cluster_metadata_complete:
+                        return self._cluster_metadata
+                elif all(topic in self._cluster_metadata["topics"] for topic in topics):
+                    return {
+                        **self._cluster_metadata,
+                        "topics": {topic: self._cluster_metadata["topics"][topic] for topic in topics},
+                    }
 
             try:
                 metadata_birth = time.monotonic()
                 metadata = self.admin_client.cluster_metadata(topics)
-                # Only store cache if fetched full metadata
-                if topics is None:
-                    self._metadata_birth = metadata_birth
-                    self._cluster_metadata = metadata
+                self._metadata_birth = metadata_birth
+                self._cluster_metadata = metadata
+                self._cluster_metadata_complete = topics is None
```

Comment:
What's the reasoning behind the following logic.

If `topics is not None`. `cluster_metadata` is called, `_cluster_metadata` is queried but only for the given topics. This value is cached but the cache is "not complete" and is not used if `cluster_metadata` is called with `topics is None`.

Imagine the following happens quickly after each other

```python
cluster_metadata() # All topics are queried and cached, image topic t is not there in the cache
cluster_metadata(topics=["t"]) # topic t is queried and cached
cluster_metadata() # All topics are queried and cached, since `_cluster_metadata_complete` is `False`
``` 

To me sounds the following would make more sense:

```python
cluster_metadata() # All topics are queried and cached, image topic t is not there in the cache
cluster_metadata(topics=["t"]) # topic t is queried, if found add that to the cached value. If not found, do not touch the cache.
cluster_metadata() # return cached value
```

I mean. If querying for a single topic, never store in the cache (bu. If querying for all topics,

Reply:
Cache is always best effort. Topics are IMHO rarely created, but one can create more than one topic at time. Your suggestion would fail in such case as "all topics" returned from cache would not contain second topic if such is created same time. Also it would make cache update logic more complex.

Reply:
I've been looking at this for a while, and it's just hard to understand:

* as said earlier, it's very odd that it sometimes caches partial results (only for the queried topics) and sometimes full results. I think it's much clearer if full results only are cached.
* I think the method implicitly raises `UnknownTopicOrPartitionError` if the topic is not found, at least that's many callers of the method handle. This is not clear in the current implementation.

Here's a more reabable idea. Basic idea: 1st check if the cache needs to be updated. If yes, update it. Then get the response from the cache. Always cache full metadata.

The downside I can think of is that is getting full metadata is more costly than getting for a single topic that doesn't exist in the cache. But if that happens the cache is fully refreshed, the follow-up calls for all topics or for that new topic are cheap because the data is in the cache. 

```python
    async def cluster_metadata(self, topic: Optional[str] = None) -> dict:
        async with self.admin_lock:
            refresh_cache = False
            if self._cluster_metadata is None:
                # First time or cache update has failed
                refresh_cache = True
            if time.monotonic() - self._metadata_birth > self.metadata_max_age:
                # cache expired
                refresh_cache = True
            elif topic is not None and topic not in self._cluster_metadata["topics"]:
                # Getting for a single topic that is not in the cache.
                # It's possible the topic has been created after the cache was populated.
                # Force cache update to not return false negative.
                refresh_cache = True

            if refresh_cache:
                self._cluster_metadata = None
                self._metadata_birth = None
                try:
                    self._cluster_metadata = self.admin_client.cluster_metadata()
                    self._metadata_birth = time.monotonic()
                except NodeNotReadyError:
                    log.exception("Could not refresh cluster metadata")
                    KafkaRest.not_found(
                        body={
                            "message": "Kafka node not ready",
                            "code": RESTErrorCodes.HTTP_INTERNAL_SERVER_ERROR.value,
                        },
                        content_type="application/json",
                        status=HTTPStatus.INTERNAL_SERVER_ERROR,
                    )

            if topic is None:
                return self._cluster_metadata
            else:
                topic_metadata = self._cluster_metadata["topics"].get(topic)
                if topic_metadata is not None:
                    return {
                        **self._cluster_metadata,
                        "topics": {topic: topic_metadata},
                    }
                else:
                    raise UnknownTopicOrPartitionError()
```

Reply:
IMHO the current implementation is shorter and more readable than your suggestion here.